### PR TITLE
[clang][dataflow] Strengthen pointer comparison.

### PR DESCRIPTION
**Pull request by martinboehme**
-  Instead of comparing the identity of the `PointerValue`s, compare the
   underlying `StorageLocation`s.

-  If the `StorageLocation`s are the same, return a definite "true" as the
   result of the comparison. Before, if the `PointerValue`s were different, we
   would return an atom, even if the storage locations themselves were the same.

-  If the `StorageLocation`s are different, return an atom (as before). Pointers
   that have different storage locations may still alias, so we can't return a
   definite "false" in this case.

The application-level gains from this are relatively modest. For the Crubit
nullability check running on an internal codebase, this change reduces the
number of functions on which the SAT solver times out from 223 to 221; the
number of "pointer expression not modeled" errors reduces from 3815 to 3778.

Still, it seems that the gain in precision is generally worthwhile.

@Xazax-hun inspired me to think about this with his
[comments](https://github.com/llvm/llvm-project/pull/73860#pullrequestreview-1761484615)
on a different PR.

### CODE REVIEW

File: clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp [hunk 1]
```diff
@@ -152,6 +153,47 @@ TEST_F(EnvironmentTest, JoinRecords) {
   }
 }
 
+TEST_F(EnvironmentTest, DifferentReferenceLocInJoin) {
```
Comment (Xazax-hun) [inline clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp:156]: While running analysis on C++ code is indirect, it has the advantage of documenting what code can trigger such behavior. I think having direct tests is fine, I just wanted to make this decision explicit.

File: clang/unittests/Analysis/FlowSensitive/TransferTest.cpp [hunk 2]
```diff
@@ -3893,6 +3893,66 @@ TEST(TransferTest, BooleanInequality) {
       });
 }
 
+TEST(TransferTest, PointerEquality) {
```
Comment (Xazax-hun) [inline clang/unittests/Analysis/FlowSensitive/TransferTest.cpp:3896]: I think it would be great to add a test case with union members to document the limitation.

**Review by Xazax-hun — APPROVED** [review]
This is a massive improvement! Thanks! Looks great to me, I have some not too important comments inline.

**Comment by Xazax-hun** [comment]
An alternative approach would be to do [hash consing](https://en.wikipedia.org/wiki/Hash_consing).  Since pointer values are completely determined by the pointee (for some definition of the pointee, e.g., we might have `void*` and `int*` pointers pointing to the same storage location), we could have a single representation for all of the equal pointer values and continue to rely on identity. This can also be beneficial for memory consumption, but we have to pay for that by making value creation a bit more expensive (and the pointer values needs to be immutable). 

That being said, this would be a bit of a departure from the current architecture, so I think even if we want to do this, it should be a separate PR.

File: clang/unittests/Analysis/FlowSensitive/TransferTest.cpp (hunk 2 above)
Comment (martinboehme) [reply clang/unittests/Analysis/FlowSensitive/TransferTest.cpp:3896]: Good point -- done.

File: clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp (hunk 1 above)
Comment (martinboehme) [reply clang/unittests/Analysis/FlowSensitive/DataflowEnvironmentTest.cpp:156]: Agree. The original test case was minimized from a failing assertion, but it would likely be fiddly to try and recreate a test case that triggers this again. Testing directly also has the benefit of being less brittle, so I think it's an acceptable tradeoff in this case.

**Comment by martinboehme** [comment]
> An alternative approach would be to do [hash consing](https://en.wikipedia.org/wiki/Hash_consing). Since pointer values are completely determined by the pointee (for some definition of the pointee, e.g., we might have `void*` and `int*` pointers pointing to the same storage location), we could have a single representation for all of the equal pointer values and continue to rely on identity. This can also be beneficial for memory consumption, but we have to pay for that by making value creation a bit more expensive (and the pointer values needs to be immutable).
> 
> That being said, this would be a bit of a departure from the current architecture, so I think even if we want to do this, it should be a separate PR.

Thanks, these are good points.

I think we may want to revisit multiple aspects of pointer values and storage locations at some point. For example, storage locations always have a type, and so storage locations of different types necessarily compare different. For example:

```cxx
struct Base {};
struct Derived : public Base {};
Derived d;
Derived *p_derived = &d;
Base *p_base = d;

// `p_derived == p_base` should hold, but the framework assumes `p_derived != p_base`.
```

It's probably a good thing for storage locations to have types (because we want to be able to model the dynamic type of an object), but this probably means that we want to model addresses separately from storage locations. This would then also help in other cases, such as union members and pointer interconvertibility. But all of this would be a major undertaking, I think.

**Comment by Xazax-hun** [comment]
> It's probably a good thing for storage locations to have types (because we want to be able to model the dynamic type of an object), but this probably means that we want to model addresses separately from storage locations.

Our knowledge of the dynamic type might change during the analysis. Consider the following snippet:

```
void f(Base* b)
{
  if (dynamic_cast<Derived*>(b)) {
    // Here, we know that the dynamic type is (a subclass of) Derived
  }
}
``` 

So, we either need to mutate storage locations, or we need to store the dynamic types on the side. The Clang Static Analyzer (CSA) is doing the latter. While CSA currently has types associated with the storage locations, we are considering undoing this decision and make storage locations only represented by sizes and offsets. We plan to always rely on the AST to determine the static types, and use a data structure on the side to track the dynamic type information. 

That being said, for some frameworks having typed storage locations might be a good idea, it is just not obvious to me where is the sweet spot and it will need careful consideration. (And as far as I understand, we are not even sure whether we want to switch to a different representation like access paths in the future.) 

> But all of this would be a major undertaking, I think.

Absolutely agree. The Clang Static Analyzer still has problems with type casts in some cases, it is a really challenging problem. That being said, the CSA is already immensely useful even with the partial modeling it has, so this is probably one of the less pressing problems to solve in the near/medium term.

**Comment by martinboehme** [comment]
Heads up: I'm holding off on merging this patch because it makes some tests in the [Crubit nullability analysis](https://github.com/google/crubit/tree/main/nullability) fail. I'm still investigating exactly why this is.

**Review by ymand — APPROVED** [review]

**Comment by martinboehme** [comment]
Finally getting back to this.

The reason this was causing tests to fail is really pretty obvious in hindsight: Different `StorageLocation`s may alias, so if the `StorageLocation`s for two `PointerValue`s are different, we can't assume that the pointers must compare unequal.

Here's an example:

```cxx
void f(int *p1, int *p2) {
  if (p1 == p2) {
    do_something();
  }
}
```

The framework initializes `p1` and `p2` with different storage locations, so the current state of this PR causes the comparison `p1 == p2` to be evaluated to a false literal, which in turn causes the `do_something()` block to be considered dead code. But this is nonsense: `p1` and `p2` may alias, so it's definitely possible that `do_something()` will be executed.

In summary, if the storage locations for two pointers are the same, we can return a true literal for the comparison, but if the storage locations are different, we need to return an atom.

I'll change the PR accordingly and will let you know once it's ready for another look.

**Comment by martinboehme** [comment]
> I'll change the PR accordingly and will let you know once it's ready for another look.

Sorry, this completely fell off my radar. I've finally revisited this now, and the PR is ready for review again. PTAL.

**Review by ymand — APPROVED** [review]

**Review by Xazax-hun — APPROVED** [review]

**Comment by Xazax-hun** [comment]
> In summary, if the storage locations for two pointers are the same, we can return a true literal for the comparison, but if the storage locations are different, we need to return an atom.

I am wondering, if it would make sense to "fix up" the state when we discover aliasing? I.e., "merging" the two storage locations. Consider:

```
void f(MyBoolPair* a, MyBoolPair *b) {
  if (a == b && a->first && !b->second)
  {
    // Here, we should know: a->first == b->first == true and a->second == b->second == false
  } 
}
```

**Comment by martinboehme** [comment]
> > In summary, if the storage locations for two pointers are the same, we can return a true literal for the comparison, but if the storage locations are different, we need to return an atom.
> 
> I am wondering, if it would make sense to "fix up" the state when we discover aliasing? I.e., "merging" the two storage locations. Consider:
> 
> ```
> void f(MyBoolPair* a, MyBoolPair *b) {
>   if (a == b && a->first && !b->second)
>   {
>     // Here, we should know: a->first == b->first == true and a->second == b->second == false
>   } 
> }
> ```

This sounds great -- but I think getting this to work wouldn't be trivial. At least, I don't yet have any good ideas on how to make it work.

I think we'd have to do something like this:

1.  When we evaluate `a == b`, associate this expression with an atom, then remember that this atom means that the storage locations for the two pointers are the same.

2.  When we enter a new block, if we can prove from the flow condition that the atom mentioned in step 1 is true, remember that the storage locations for `a` and `b` must alias.

3.  When we evaluate `b->first` -- I guess we iterate over all of the storage locations that must alias with the storage location for `b`, and if, for at least one of them, we can prove that `b->first` is true, we know it must be true? (I'm not super-conversant in alias analysis -- would have to read up more about this.)

Would require some more thought and definitely quite a bit of effort to implement, I think.

**Comment by martinboehme** [comment]
CI failure (DataFlowSanitizer-x86_64 :: release_shadow_space.c) looks unrelated.
